# Jupyter notebook info

The first thing we need to know about jupyter notebook is that there are two kinds of cells. This is a markdown cell.

To change to the other kind of cell, called **code** cell enter in command mode, pressing ESC and type *y*, to back to markdown type *m*.

To execute a cell content press SHIFT + ENTER

# Credit card clients project details

## Introduction
This is my first data science project, where I've begun organizing and applying the knowledge I've acquired. It's not a large-scale project, but the idea is to take baby steps, gradually increasing the difficulty level until tackling more complex projects.

I should clarify that while I'm not the original author of this project (including the dataset and code), I'm following a data science course. However, throughout this project, I've made some adjustments and incorporated notes based on what I've learned.

## Dataset

The dataset for this project was sourced from the UCI Machine Learning Repository [http://archive.ics.uci.edu/ml], provided by the University of California, School of Information and Computer Science. It contains demographic and financial data from the latest 6 months and comprises around 30,000 samples.

## Project goal

To accomplish this project, we will explore the dataset, gaining important insights and understanding its contents. Additionally, we will clean the dataset using Pandas and apply various data science techniques.

## Data science techniques

Boolean masks, 

## Vocabulary
 - Dataset -> the whole data
 - Row definition -> what each row of dataset means
 - Labels -> column names from dataset
 - Data dictionary -> document that explains the definitions of column labels
 - Dataframe -> python object representing the dataset
 - Series -> python object representing the column from dataframe

## Study case

In this project, we will examine a case from a credit card company. They have provided us with a dataset containing the last 6 months of client accounts at the account level. In other words, there **is one row for each account**. (*It's essential to always define the row definition in a dataset*).

The rows are labeled based on whether a client defaulted in the following month relative to the historical period.

## Challenge

Our goal is develop a model that can predict whether an account will default on its bills in the next month, based on demographic and financial data.

#### Step 1 - Load the dataset and getting shape info (rows and columns)

In [2]:
import pandas as pd

In [3]:
df = pd.read_excel('credit_card_clients.xls')

In [4]:
df.shape

(30000, 25)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   ID                          30000 non-null  object
 1   LIMIT_BAL                   30000 non-null  int64 
 2   SEX                         30000 non-null  int64 
 3   EDUCATION                   30000 non-null  int64 
 4   MARRIAGE                    30000 non-null  int64 
 5   AGE                         30000 non-null  int64 
 6   PAY_1                       30000 non-null  object
 7   PAY_2                       30000 non-null  int64 
 8   PAY_3                       30000 non-null  int64 
 9   PAY_4                       30000 non-null  int64 
 10  PAY_5                       30000 non-null  int64 
 11  PAY_6                       30000 non-null  int64 
 12  BILL_AMT1                   30000 non-null  int64 
 13  BILL_AMT2                   30000 non-null  in

In [6]:
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0  798fc410-45c1      20000    2          2         1   24     2      2   
1  8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2  85698822-43f5      90000    2          2         2   34     0      0   
3  0737c11b-be42      50000    2          2         1   37     0      0   
4  3b7f77cc-dbc0      50000    1          2         1   57    -1      0   

   PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0     -1     -1  ...          0          0          0         0       689   
1      0      0  ...       3272       3455       3261         0      1000   
2      0      0  ...      14331      14948      15549      1518      1500   
3      0      0  ...      28314      28959      29547      2000      2019   
4     -1      0  ...      20940      19146      19131      2000     36681   

   PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0         0         0         0         0                           1  
1      1000      1000         0      2000                           1  
2      1000      1000      1000      5000                           0  
3      1200      1100      1069      1000                           0  
4     10000      9000       689       679                           0  

[5 rows x 25 columns]

Above we show the `shape` to get the rows and columns quantity

after we executed `info` that brings to us each column and datatype

And `head`, that show the first rows from the dataset

#### Step 2 - Exploring data

Generally, there are two kinds of dataset sources: those **you participated** in constructing and those you **did not**.

To aid in the exploration of data, we should utilize a **data dictionary**, which is a document that explains the definitions of column labels.

If you created the dataset, it is your responsibility to generate the **data dictionary**.

If you did not create the dataset, you will need to attempt to obtain the **data dictionary** whenever possible.

#### Dataset dictionary
This research employed a binary variable, default payment (**Yes = 1, No = 0**), as the response variable. This study reviewed the literature and used the following 23 variables as explanatory variables:

**LIMIT_BAL**: Amount of the given credit (NT dollar): it includes both the individual consumer credit and his/her family (supplementary) credit.

**SEX**: Gender (**1 = male; 2 = female**).

**EDUCATION**: Education (**1 = graduate school; 2 = university; 3 = high school; 4 = others**).

**MARRIAGE**: Marital status (**1 = married; 2 = single; 3 = others**).

**AGE**: Age (year).

**PAY_1 - PAY_6**: History of past payment. We tracked the past monthly payment records (from April to September, 2005) as follows: PAY_1 = the repayment status in September, 2005; PAY_2 = the repayment status in August, 2005; . . .;PAY_6 = the repayment status in April, 2005. The measurement scale for the repayment status is: **-1 = pay duly; 1 = payment delay for one month; 2 = payment delay for two months; . . .; 8 = payment delay for eight months; 9 = payment delay for nine months and above.**

**BILL_AMT1 - BILL_AMT6**: Amount of bill statement (NT dollar). BILL_AMT1 = amount of bill statement in September, 2005; BILL_AMT2 = amount of bill statement in August, 2005; . . .; BILL_AMT6 = amount of bill statement in April, 2005. 

**PAY_AMT1 - PAY_AMT6**: Amount of previous payment (NT dollar). PAY_AMT1 = amount paid in September, 2005; PAY_AMT2 = amount paid in August, 2005; . . .;PAY_AMT6 = amount paid in April, 2005.

#### Exploring data - basic steps

We have the follow steps to exploring data:

    1 - Knows how many columns has the dataset (can be: characteristics, answers or metadata)

    2 - How many rows (samples)

    3 - What kind of characteristics exists? What are categories and what are numerics?

    4 - Which is the data apparence according with its charecteristics
        - For example: interval between values on numerics charecteristics or frequency of class in category charecteristics
        
    5 - Is there null data?

Let´s pratice it!

##### Exploring data - Question 1 / Question 2

In [7]:
df.shape

(30000, 25)

Now let´s check if there is duplicate data on our dataset, to do that, we will use the id column, because is supposed to be unique

In [8]:
df.columns

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_1',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default payment next month'],
      dtype='object')

In [10]:
df['ID'].nunique()

29687

Explaing the command above, so fist we select a pandas column and got a Series with: *df['ID']* that is: dataframe + [' + column_name ' +]

So on, we apply the nunique() function

We figure out our first data problem, nunique function shows 29687 while our dataset has 30000 rows, this means that we have duplicate data

Now, we need to investigate it and answer some question like:
- May does only one ID is duplicate many times?
- How many IDs are duplicate?

To answer, we will use the value_counts function. This is similar to count/group by from SQL

In [11]:
id_counts = df['ID'].value_counts()

In [12]:
id_counts.head()

ID
ad23fe5c-7b09    2
1fb3e3e6-a68d    2
89f8f447-fca8    2
7c9b7473-cc2f    2
90330d02-82d9    2
Name: count, dtype: int64

**head function returned the first 5 results. If needed you can modify the results quantity, passing head(number)**



In [14]:
id_counts.head(10)

ID
ad23fe5c-7b09    2
1fb3e3e6-a68d    2
89f8f447-fca8    2
7c9b7473-cc2f    2
90330d02-82d9    2
2a793ecf-05c6    2
75938fec-e5ec    2
7be61027-a493    2
a3a5c0fc-fdd6    2
b44b81b2-7789    2
Name: count, dtype: int64

Now we will show the duplicated IDs grouping by values

In [15]:
id_counts.value_counts()

count
1    29374
2      313
Name: count, dtype: int64

We can observe that there are 29,374 IDs with one occurrence each, and 313 IDs with two occurrences each. This means that 313 IDs appear twice, and no ID appears three times